python/pylibraft/pylibraft/neighbors/refine.pyx

```diff
@@ -0,0 +1,461 @@
+        get_device_matrix_view_uint64(indices)
+
+    cdef device_matrix_view[float, uint64_t, row_major] distances_view = \
+        get_device_matrix_view_float(distances)
+
+    cdef DistanceType c_metric = _get_metric(metric)
+
+    dataset_cai = cai_wrapper(dataset)
+
+    if dataset_cai.dtype == np.float32:
+        dataset_view_fp32 = get_device_matrix_view_float(dataset)
+        queries_view_fp32 = get_device_matrix_view_float(queries)
+        with cuda_interruptible():
+            c_refine(deref(handle_),
+                     dataset_view_fp32,
+                     queries_view_fp32,
+                     candidates_view,
+                     indices_view,
+                     distances_view,
+                     c_metric)
+    elif dataset_cai.dtype == np.int8:
+        dataset_view_int8 = get_device_matrix_view_int8(dataset)
+        queries_view_int8 = get_device_matrix_view_int8(queries)
+        with cuda_interruptible():
+            c_refine(deref(handle_),
+                     dataset_view_int8,
+                     queries_view_int8,
+                     candidates_view,
+                     indices_view,
+                     distances_view,
+                     c_metric)
+    elif dataset_cai.dtype == np.uint8:
+        dataset_view_uint8 = get_device_matrix_view_uint8(dataset)
+        queries_view_uint8 = get_device_matrix_view_uint8(queries)
+        with cuda_interruptible():
+            c_refine(deref(handle_),
+                     dataset_view_uint8,
+                     queries_view_uint8,
+                     candidates_view,
+                     indices_view,
+                     distances_view,
+                     c_metric)
+    else:
+        raise TypeError("dtype %s not supported" % dataset_cai.dtype)
+
+    return (distances, indices)
+
+
+def _refine_host(dataset, queries, candidates, k, indices, distances,
+                 metric, handle):
+    cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
+
+    cdef host_matrix_view[float, uint64_t, row_major] dataset_view_fp32
+    cdef host_matrix_view[uint8_t, uint64_t, row_major] dataset_view_uint8
+    cdef host_matrix_view[int8_t, uint64_t, row_major] dataset_view_int8
+    cdef host_matrix_view[float, uint64_t, row_major] queries_view_fp32
+    cdef host_matrix_view[uint8_t, uint64_t, row_major] queries_view_uint8
+    cdef host_matrix_view[int8_t, uint64_t, row_major] queries_view_int8
+
+    cdef host_matrix_view[uint64_t, uint64_t, row_major] candidates_view = \
+        get_host_matrix_view_uint64(candidates)
+
+    if k is None:
+        if indices is not None:
+            k = indices.__array_interface__["shape"][1]
+        elif distances is not None:
+            k = distances.__array_interface__["shape"][1]
+        else:
+            raise ValueError("Argument k must be specified if both indices "
+                             "and distances arg is None")
+
+    n_queries = queries.__array_interface__["shape"][0]
```

Comment:
it would be nice to see a corresponding wrapper for host functions (e.g. `ai_wrapper` maybe?) just to keep the APIs similar between device and host. We don't need to do that in this PR, but it would certainly clean this up a tad and minimize potential errors from things like string typos.

Reply:
That is a good point. I will submit a follow up PR for this.